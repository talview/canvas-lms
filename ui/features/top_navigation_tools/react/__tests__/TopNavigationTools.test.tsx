/*
 * Copyright (C) 2024 - present Instructure, Inc.
 *
 * This file is part of Canvas.
 *
 * Canvas is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * Canvas is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react'
import {shallow} from 'enzyme'
import {
  TopNavigationTools,
  MobileTopNavigationTools,
  handleToolIconError,
} from '../TopNavigationTools'
import type {Tool} from '@canvas/global/env/EnvCommon'

describe('TopNavigationTools', () => {
  it('renders', () => {
    const tools = [
      {
        id: '1',
        title: 'Tool 1',
        icon_url: 'https://instructure.com',
        pinned: true,
      },
      {
        id: '2',
        title: 'Tool 2',
        pinned: false,
      },
    ]
    const handleToolLaunch = jest.fn()
    const wrapper = shallow(
      // @ts-expect-error
      <TopNavigationTools tools={tools} handleToolLaunch={handleToolLaunch} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders with no tools', () => {
    // @ts-expect-error
    const tools = []
    const handleToolLaunch = jest.fn()
    const wrapper = shallow(
      // @ts-expect-error
      <TopNavigationTools tools={tools} handleToolLaunch={handleToolLaunch} />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders with no pinned tools', () => {
    const tools = [
      {
        id: '1',
        title: 'Tool 1',
        icon_url: 'https://instructure.com',
        pinned: false,
      },
      {
        id: '2',
        title: 'Tool 2',
        icon_url: 'https://instructure.com',
        pinned: false,
      },
    ]
    const handleToolLaunch = jest.fn()
    const wrapper = shallow(
      // @ts-expect-error
      <TopNavigationTools tools={tools} handleToolLaunch={handleToolLaunch} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('MobileTopNavigationTools', () => {
  it('renders', () => {
    const tools = [
      {
        id: '1',
        title: 'Tool 1',
        icon_url: 'https://instructure.com',
        pinned: false,
      },
      {
        id: '2',
        title: 'Tool 2',
        icon_url: 'https://instructure.com',
        pinned: true,
      },
    ]
    const handleToolLaunch = jest.fn()
    const wrapper = shallow(
      // @ts-expect-error
      <MobileTopNavigationTools tools={tools} handleToolLaunch={handleToolLaunch} />
    )
    expect(wrapper).toMatchSnapshot()
  })
})

describe('handleToolClick', () => {
  it('finds tool', () => {
    const tool = {
      id: '1',
      title: 'Tool 1',
      icon_url: 'https://instructure.com',
      pinned: true,
    }
    const handleToolLaunch = jest.fn()
    const wrapper = shallow(
      // @ts-expect-error
      <TopNavigationTools tools={[tool]} handleToolLaunch={handleToolLaunch} />
    )
    wrapper.find('IconButton').simulate('click', {target: {dataset: {toolId: '1'}}})
    expect(handleToolLaunch).toHaveBeenCalledWith(tool)
  })
})

describe('handleToolIconError', () => {
  it('uses default tool icon', () => {
    const tool = {
      id: '1',
      title: 'Tool 1',
      pinned: true,
    }
    const event = {
      target: {
        src: '',
      },
    }

    // @ts-expect-error
    handleToolIconError(tool)(event)

    expect(event.target.src).toBe('/lti/tool_default_icon?name=T')
  })
})
