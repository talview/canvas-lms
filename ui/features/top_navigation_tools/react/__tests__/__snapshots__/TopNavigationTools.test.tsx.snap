// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileTopNavigationTools renders 1`] = `
<Menu
  constrain="window"
  defaultShow={false}
  disabled={false}
  key="menu"
  label={null}
  mountNode={null}
  offsetX={0}
  offsetY={0}
  placement="bottom end"
  shouldFocusTriggerOnClose={true}
  shouldHideOnSelect={true}
  trigger={
    <IconButton
      as="button"
      color="secondary"
      cursor="pointer"
      margin="0"
      renderIcon={[Function]}
      screenReaderLabel="LTI Tool Menu"
      shape="rectangle"
      size="medium"
      type="button"
      withBackground={false}
      withBorder={false}
    />
  }
  withArrow={true}
>
  <MenuItem
    disabled={false}
    key="2"
    label="Launch Tool 2"
    onSelect={[Function]}
    type="button"
    value="2"
  >
    <Flex
      as="span"
      direction="row"
      display="flex"
      gap="small"
      justifyItems="start"
      withVisualDebug={false}
      wrap="no-wrap"
    >
      <Img
        alt="Tool 2"
        display="inline-block"
        height="1rem"
        onError={[Function]}
        src="https://instructure.com"
        withBlur={false}
        withGrayscale={false}
      />
      <TruncateText
        debounce={0}
        ellipsis="…"
        ignore={
          [
            " ",
            ".",
            ",",
          ]
        }
        maxLines={1}
        position="end"
        truncate="character"
      >
        Tool 2
      </TruncateText>
    </Flex>
  </MenuItem>
  <MenuItemSeparator />
  <MenuItem
    disabled={false}
    key="1"
    label="Launch Tool 1"
    onSelect={[Function]}
    type="button"
    value="1"
  >
    <Flex
      as="span"
      direction="row"
      display="flex"
      gap="small"
      justifyItems="start"
      withVisualDebug={false}
      wrap="no-wrap"
    >
      <Img
        alt="Tool 1"
        display="inline-block"
        height="1rem"
        onError={[Function]}
        src="https://instructure.com"
        withBlur={false}
        withGrayscale={false}
      />
      <TruncateText
        debounce={0}
        ellipsis="…"
        ignore={
          [
            " ",
            ".",
            ",",
          ]
        }
        maxLines={1}
        position="end"
        truncate="character"
      >
        Tool 1
      </TruncateText>
    </Flex>
  </MenuItem>
</Menu>
`;

exports[`TopNavigationTools renders 1`] = `
<Flex
  as="div"
  direction="row-reverse"
  display="flex"
  gap="small"
  height="100%"
  justifyItems="start"
  width="100%"
  withVisualDebug={false}
  wrap="no-wrap"
>
  <Item
    as="span"
    shouldGrow={false}
    shouldShrink={false}
  >
    <Menu
      constrain="window"
      defaultShow={false}
      disabled={false}
      key="menu"
      label="LTI Tools Menu"
      mountNode={null}
      offsetX={0}
      offsetY={0}
      placement="bottom end"
      shouldFocusTriggerOnClose={true}
      shouldHideOnSelect={true}
      trigger={
        <Tooltip
          color="primary"
          constrain="window"
          defaultIsShowingContent={false}
          offsetX={0}
          offsetY={0}
          placement="top"
          renderTip="LTI Tools Menu"
        >
          <Button
            as="button"
            color="secondary"
            cursor="pointer"
            display="inline-block"
            margin="0"
            renderIcon={[Function]}
            size="medium"
            textAlign="center"
            type="button"
            withBackground={true}
          />
        </Tooltip>
      }
      withArrow={true}
    >
      <MenuItem
        disabled={false}
        key="2"
        label="Launch Tool 2"
        onSelect={[Function]}
        type="button"
        value="2"
      >
        <Flex
          as="span"
          direction="row"
          display="flex"
          gap="small"
          justifyItems="start"
          withVisualDebug={false}
          wrap="no-wrap"
        >
          <Img
            alt="Tool 2"
            display="inline-block"
            height="1rem"
            onError={[Function]}
            src=""
            withBlur={false}
            withGrayscale={false}
          />
          <TruncateText
            debounce={0}
            ellipsis="…"
            ignore={
              [
                " ",
                ".",
                ",",
              ]
            }
            maxLines={1}
            position="end"
            truncate="character"
          >
            Tool 2
          </TruncateText>
        </Flex>
      </MenuItem>
    </Menu>
  </Item>
  <Item
    as="span"
    key="1"
    shouldGrow={false}
    shouldShrink={false}
  >
    <Tooltip
      color="primary"
      constrain="window"
      defaultIsShowingContent={false}
      offsetX={0}
      offsetY={0}
      placement="top"
      renderTip="Tool 1"
    >
      <IconButton
        as="button"
        color="secondary"
        cursor="pointer"
        data-tool-id="1"
        margin="0"
        onClick={[Function]}
        renderIcon={
          <Img
            alt="Tool 1"
            display="inline-block"
            height="1rem"
            onError={[Function]}
            src="https://instructure.com"
            withBlur={false}
            withGrayscale={false}
          />
        }
        screenReaderLabel="Tool 1"
        shape="rectangle"
        size="medium"
        type="button"
        withBackground={true}
        withBorder={true}
      />
    </Tooltip>
  </Item>
</Flex>
`;

exports[`TopNavigationTools renders with no pinned tools 1`] = `
<Flex
  as="div"
  direction="row-reverse"
  display="flex"
  gap="small"
  height="100%"
  justifyItems="start"
  width="100%"
  withVisualDebug={false}
  wrap="no-wrap"
>
  <Item
    as="span"
    shouldGrow={false}
    shouldShrink={false}
  >
    <Menu
      constrain="window"
      defaultShow={false}
      disabled={false}
      key="menu"
      label="LTI Tools Menu"
      mountNode={null}
      offsetX={0}
      offsetY={0}
      placement="bottom end"
      shouldFocusTriggerOnClose={true}
      shouldHideOnSelect={true}
      trigger={
        <Tooltip
          color="primary"
          constrain="window"
          defaultIsShowingContent={false}
          offsetX={0}
          offsetY={0}
          placement="top"
          renderTip="LTI Tools Menu"
        >
          <Button
            as="button"
            color="secondary"
            cursor="pointer"
            display="inline-block"
            margin="0"
            renderIcon={[Function]}
            size="medium"
            textAlign="center"
            type="button"
            withBackground={true}
          />
        </Tooltip>
      }
      withArrow={true}
    >
      <MenuItem
        disabled={false}
        key="1"
        label="Launch Tool 1"
        onSelect={[Function]}
        type="button"
        value="1"
      >
        <Flex
          as="span"
          direction="row"
          display="flex"
          gap="small"
          justifyItems="start"
          withVisualDebug={false}
          wrap="no-wrap"
        >
          <Img
            alt="Tool 1"
            display="inline-block"
            height="1rem"
            onError={[Function]}
            src="https://instructure.com"
            withBlur={false}
            withGrayscale={false}
          />
          <TruncateText
            debounce={0}
            ellipsis="…"
            ignore={
              [
                " ",
                ".",
                ",",
              ]
            }
            maxLines={1}
            position="end"
            truncate="character"
          >
            Tool 1
          </TruncateText>
        </Flex>
      </MenuItem>
      <MenuItem
        disabled={false}
        key="2"
        label="Launch Tool 2"
        onSelect={[Function]}
        type="button"
        value="2"
      >
        <Flex
          as="span"
          direction="row"
          display="flex"
          gap="small"
          justifyItems="start"
          withVisualDebug={false}
          wrap="no-wrap"
        >
          <Img
            alt="Tool 2"
            display="inline-block"
            height="1rem"
            onError={[Function]}
            src="https://instructure.com"
            withBlur={false}
            withGrayscale={false}
          />
          <TruncateText
            debounce={0}
            ellipsis="…"
            ignore={
              [
                " ",
                ".",
                ",",
              ]
            }
            maxLines={1}
            position="end"
            truncate="character"
          >
            Tool 2
          </TruncateText>
        </Flex>
      </MenuItem>
    </Menu>
  </Item>
</Flex>
`;

exports[`TopNavigationTools renders with no tools 1`] = `
<Flex
  as="div"
  direction="row-reverse"
  display="flex"
  gap="small"
  height="100%"
  justifyItems="start"
  width="100%"
  withVisualDebug={false}
  wrap="no-wrap"
/>
`;
